packages/grafana-ui/src/components/Table/TableCell.tsx

```diff
@@ -1,38 +1,59 @@
 import React, { FC } from 'react';
 import { Cell } from 'react-table';
 import { Field } from '@grafana/data';
+
 import { getTextAlign } from './utils';
-import { TableFilterActionCallback } from './types';
+import { TableCellClickActionCallback, TableFilterActionCallback } from './types';
 import { TableStyles } from './styles';
+import { ClickableTableCell } from './ClickableTableCell';
+import { FilterableTableCell } from './FilterableTableCell';
 
-interface Props {
+export interface Props {
   cell: Cell;
   field: Field;
   tableStyles: TableStyles;
-  onCellClick?: TableFilterActionCallback;
+  onCellClick?: TableCellClickActionCallback;
+  onFilterAdded?: TableFilterActionCallback;
 }
 
-export const TableCell: FC<Props> = ({ cell, field, tableStyles, onCellClick }) => {
+export const TableCell: FC<Props> = ({ cell, field, tableStyles, onCellClick, onFilterAdded }) => {
   const filterable = field.config.filterable;
   const cellProps = cell.getCellProps();
 
-  let onClick: ((event: React.SyntheticEvent) => void) | undefined = undefined;
+  if (cellProps.style) {
+    cellProps.style.textAlign = getTextAlign(field);
+  }
 
   if (filterable && onCellClick) {
-    if (cellProps.style) {
-      cellProps.style.cursor = 'pointer';
-    }
-
-    onClick = () => onCellClick(cell.column.Header as string, cell.value);
+    return (
+      <ClickableTableCell
+        cell={cell}
+        field={field}
+        tableStyles={tableStyles}
+        onCellClick={onCellClick}
+        cellProps={cellProps}
+      />
+    );
   }
 
-  if (cellProps.style) {
-    cellProps.style.textAlign = getTextAlign(field);
+  if (filterable && onFilterAdded) {
+    return (
+      <FilterableTableCell
+        cell={cell}
+        field={field}
+        tableStyles={tableStyles}
+        onFilterAdded={onFilterAdded}
+        cellProps={cellProps}
+      />
+    );
```

Comment:
Would love to get feedback on this refactor 🙏

Reply:
This sort of relates to my previous comment on how we enable the filtering callback. Since we want to inject domain-specific props to the table cell I think we would have to introduce a render prop for the cell. Then not sure we have/will have a case for filterable prop in near time feature, hence introducing this additional prop feels like introducing unnecessary complexity. WDYT?

Reply:
not sure what the best solution is, we need this in Explore as well. filterable is a field.config property. think a prop for this on Table is ok. wonder if we should name the callback a bit differently, like onCellFilterAdded , as I think we will add filters on columns as well in the future (think filters in excel on the headers), to filter values in a table in the table component it'self.

Reply:
I like this suggestion; it makes the api clearer.

Reply:
Yeah, I like the suggestion as well. Think as long as the props are all optional this makes sense.